### DISCUSSION TITLE
Wallet Activity Notification Feature

### DISCUSSION THREAD

**Discussion by pxsocs** [cryptoadvance/specter-desktop#1239] [Specter UI/UX]
@ben-kaufman @k9ert 

Looking into creating a wallet activity notification feature. Wanted to get your guidance on best way to implement this. This is something I've implemented in the WARden and think could be a useful feature for Specter. Quick overview on how this could work. 

1. When application launches, start a background job that checks for activity every [n] seconds
    . Does the app already have any background jobs method that could be reused / based off?
    Suggestion would be to do something along these lines:
    ```python
    def check_activity_bk():
        # include check here
        pass

    from apscheduler.schedulers.background import BackgroundScheduler
    app = Flask(__name__)
    app.scheduler = BackgroundScheduler()
    app.scheduler.add_job(check_activity_bk, 'interval', seconds=10)
    ```
    
2. A dictionary is saved as a pickle on default home folder (i.e. ~/specter ?):
```json
{'wallet_alias_1' : 
    {
    'last_tx_hash' : '------',
    'last_balance_hash': '----',
     }, 
 'wallet_alias_2': ... 
}
```

3. When checking for activity, the latest hashes are checked against the pkl file. If changes detected, an icon appears by the wallet name on the menu bar: 'Activity Detected since last checkpoint'. A 'dismiss' button would save the pkl file with the latest checkpoint and stop notifying the user. Notification will keep being displayed until 'dismiss' is clicked.  

4. Include in the `specter` class, a new dictionary: `wallet_activity` so it's easily accessible across html jinja2 templates:
```json
{
    'activity_detected': False,
    'latest_status' : {...}, 
    'pkl_status': {...}
}
```
`latest_status` and `pkl_status` are dictionaries in the format listed at (2). Saving both will make it easy to check which wallets have balance change and/or txs. 

If this is acceptable and you believe this to be a useful feature, I'm happy to work on some initial code and submit a PR. 

This could also be enhanced later to send e-mail notifications, etc. But would need to consider potential risks.

**Comment by k9ert** [comment]
Well, that sounds cool :-).

> When application launches, start a background job that checks for activity every [n] seconds

That's one possibility but Bitcoin-core has a better one: ZeroMQ. We can have nearly instant notification and don't have to rely on pulling with ZMQ.
Have a look e.g. here:
https://bitcoindev.network/accessing-bitcoins-zeromq-interface/
This is based on nodejs. Here is a python-example:
https://github.com/bitcoin/bitcoin/blob/master/contrib/zmq/zmq_sub.py
 
> A dictionary is saved as a pickle on default home folder (i.e. ~/specter ?):

We should imho store wallet-specific stuff in the wallet-folder. Check `~/.specter/wallets/main/...`

> Include in the specter class, a new dictionary: wallet_activity so it's easily accessible across html jinja2 templates:

I'd suggest a new Manager similiar to WalletManager, UserManager and so on. Check `src/cryptoadvance/specter/managers`. Initialize them in the __init__-method of specter similiar to the other managers. Maybe you can call it `ZmqManager` so that it receives all events from BitcoinCore (and later maybe Liquid) coming via ZMQ. There are then one method which is storing the pickles and a method which checks for activity and reads the corresponding pickles and returns the result.

> If this is acceptable and you believe this to be a useful feature, I'm happy to work on some initial code and submit a PR.

Looking forward to your PR. In the meantime i'll try to get the API ready :-).

But maybe let's wait also on comments from @stepansnigirev  and @ben-kaufman
